src/mlpack/tests/string_encoding_test.cpp

```diff
@@ -0,0 +1,640 @@
+    "ABCABCD",
+    "GAB"
+  };
+
+  vector<vector<size_t>> output;
+  BowEncoding<CharExtract::TokenType> encoder;
+
+  // Passing a empty string to encode characters
+  encoder.Encode(input, output, CharExtract());
+
+  vector<vector<size_t>> expected = {
+    { 1, 1, 1, 0, 0 },
+    { 0, 1, 1, 1, 1 },
+    { 1, 1, 0, 1, 0 }
+  };
+
+  BOOST_REQUIRE(output == expected);
+}
+
+BOOST_AUTO_TEST_CASE(TfIdfEncodingTest)
+{
+  using DictionaryType = StringEncodingDictionary<boost::string_view>;
+
+  arma::mat output;
+  TfIdfEncoding<SplitByAnyOf::TokenType> encoder;
+  SplitByAnyOf tokenizer(" ");
+
+  encoder.Encode(stringEncodingInputSmall, output, tokenizer);
+
+  const DictionaryType& dictionary = encoder.Dictionary();
+
+  // Checking that everything is mapped to different numbers
+  std::unordered_map<size_t, size_t> keysCount;
+  for (auto& keyValue : dictionary.Mapping())
+  {
+    keysCount[keyValue.second]++;
+    // Every token should be mapped only once
+    BOOST_REQUIRE_EQUAL(keysCount[keyValue.second], 1);
+  }
+  arma::mat expected = {
+    { 0.1193, 0.0440, 0.0440, 0.0440, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0.1590, 0.1590, 0.1590, 0 },
+    { 0, 0.0440, 0.0440, 0.0440, 0, 0, 0, 0.1193 }
+  };
+  CheckMatrices(output, expected, 1e-01);
+}
+
+/**
+ * Test the one pass modification of the TfIdf encoding algorithm.
+ */
+BOOST_AUTO_TEST_CASE(OnePassTfIdfEncodingTest)
+{
+  using DictionaryType = StringEncodingDictionary<boost::string_view>;
+
+  vector<vector<double>> output;
+  TfIdfEncoding<SplitByAnyOf::TokenType> encoder(
+      (TfIdfEncodingPolicy()));
+  SplitByAnyOf tokenizer(" ");
+
+  encoder.Encode(stringEncodingInputSmall, output, tokenizer);
+
+  const DictionaryType& dictionary = encoder.Dictionary();
+
+  // Checking that everything is mapped to different numbers
+  std::unordered_map<size_t, size_t> keysCount;
+  for (auto& keyValue : dictionary.Mapping())
+  {
+    keysCount[keyValue.second]++;
+    // Every token should be mapped only once
+    BOOST_REQUIRE_EQUAL(keysCount[keyValue.second], 1);
+  }
+
+  vector<vector<double>> expected = {
+    { 0.11928, 0.0440228, 0.0440228, 0.0440228, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0.15904, 0.15904, 0.15904, 0 },
+    { 0, 0.044028, 0.0440228, 0.0440228, 0, 0, 0, 0.11928 }
+  };
+  for(size_t i=0;i<expected.size();i++)
+    for(size_t j=0;j<expected[i].size();j++)
+      BOOST_REQUIRE_CLOSE(expected[i][j], output[i][j], 1e-01);
+}
+
+/**
+* Test TFIDF encoding for characters using lamda function.
+*/
+BOOST_AUTO_TEST_CASE(TfIdfEncodingIndividualCharactersTest)
+{
+  vector<string> input = {
+    "GACCA",
+    "ABCABCD",
+    "GAB"
+  };
+
+  arma::mat output;
+  TfIdfEncoding<CharExtract::TokenType> encoder;
+
+  // Passing a empty string to encode characters
+  encoder.Encode(input, output, CharExtract());
+  arma::mat target = {
+    { 0.0352, 0, 0.0704, 0, 0 },
+    { 0, 0, 0.0503, 0.0503, 0.0682 },
+    { 0.0587, 0, 0, 0.0587, 0 }
+  };
+  CheckMatrices(output, target, 1e-01);
```

Comment:
`1e-12`.

Reply:
also isn't 1e-12 too strict ?
I have to type 12 digits after decimal

can't we make it
1e-10 ?

Reply:
I don't see any difference between `1e-10` and `1e-12`. There is no need to type it. You can write the result of `scikit-learn` to a file using the desired format.

Reply:
Ok , I have used the following [article](https://towardsdatascience.com/tfidf-for-piece-of-text-in-python-43feccaa74f8) to write the implemetation